### PR TITLE
Strip registry root from `file` in `Index::add()` unconditionally

### PR DESCRIPTION
**Pull request by twistedfall**
Hi!

That PR removes the `.is_relative()` check from `Index::add()`. The cause for this is that in my tests when started with the relative path as registry root (e.g. `cargo-http-registry crates-local` publishing crates to such registry fails with:
```
error: failed to publish to registry at http://127.0.0.1:34245

Caused by:
  the remote server responded with an error: failed to add crates-local/3/t/ttt to git repository, failed to add file to git index, could not find '/home/user/projects/crates-local/crates-local/3/t/ttt' to stat: No such file or directory; class=Os (2); code=NotFound (-3)
```
So the `crates-local` is duplicated. This is caused by the fact that when constructing `crate_meta_path` the base is index root (`crates-local`) which makes the whole `crate_meta_path` also relative. Then adding it fails because the stripping of the index root is not performed.

An alternative solution could probably be to avoid calling `Index::add` with the paths that contain index root, but I haven't investigated this too much.

### CODE REVIEW

**Comment by d-e-s-o** [comment]
Hi, thanks for the pull request! Can you please add a test case for the problem being fixed?

**Comment by twistedfall** [comment]
I'll do that, but I have since thought of a corner case that this PR actually breaks. If your registry root dir is called `config.json` and you use the relative path when you run the server `cargo-http-registry config.json`, which breaks of course.

**Comment by twistedfall** [comment]
I have changed the approach to only pass relative paths to `Index::add()` and added the tests for the use-case of passing relative registry root dir

File: tests/end-to-end.rs [hunk 1]
```diff
@@ -161,28 +167,54 @@ where
...
 
 /// Serve our registry.
-fn serve_registry() -> (JoinHandle<()>, PathBuf, SocketAddr) {
-  let root = tempdir().unwrap();
-  let path = root.path().to_owned();
+fn serve_registry(root_path: RegistryRootPath) -> (JoinHandle<()>, PathBuf, SocketAddr) {
+  let path = match root_path {
+    RegistryRootPath::Absolute => tempdir().unwrap().path().to_owned(),
+    RegistryRootPath::Relative => {
+      let root = tempdir().unwrap().into_path();
```
Comment (d-e-s-o) [inline tests/end-to-end.rs]: So now we are leaking temporary directories all over the place. Not okay.

File: tests/end-to-end.rs [hunk 2]
```diff
@@ -161,28 +167,54 @@ where
...
+      let base_dir = root.parent().unwrap();
+      env::set_current_dir(base_dir).unwrap();
```
Comment (d-e-s-o) [inline tests/end-to-end.rs]: Why do we need to change the current working directory everywhere? This is a nightmare in a multi-threaded context. Plus, if I remove the call everything is fine...

File: Cargo.toml [hunk 3]
```diff
@@ -32,10 +32,11 @@ opt-level = "z"
 lto = true
 codegen-units = 1
 incremental = false
+strip = true
```
Comment (d-e-s-o) [inline Cargo.toml]: No unrelated changes please.

Comment (twistedfall) [reply Cargo.toml]: Got it, reverted

File: tests/end-to-end.rs (hunk 1 above)
Comment (twistedfall) [reply tests/end-to-end.rs]: Good catch, it's now fixed

File: tests/end-to-end.rs (hunk 2 above)
Comment (twistedfall) [reply tests/end-to-end.rs]: This is needed to make passing relative directory to `Index::new()` work inside the temporary directory. If you remove this then it the index root will be re-created by the `Index::new()` in the directory where you run the test command, i.e. project root dir most likely

Comment (d-e-s-o) [reply tests/end-to-end.rs]: I don't get it. I don't see anything in `Index::new` relying on the current working directory. If there is something, that's probably a bug. But sprinkling `set_current_dir` here is not a fix. Again, tests run concurrently and will trample over each other's `set_current_dir` calls.

Comment (twistedfall) [reply tests/end-to-end.rs]: As soon as you pass a relative path in `root` argument to `Index::new()` then everything starts depending on the current directory :) E.g. if we pass a literal `"index-root"` as `root` argument then for example `Index::ensure_config()` will write `config.json` to `"index-root/config.json"` which effectively writes to `"<CWD>/index-root/config.json"`.

I understand your concerns, but I don't really see an easier way to test that relative paths work while still remaining constrained to the temporary directory.

Comment (twistedfall) [reply tests/end-to-end.rs]: Ultimately I don't mind dropping this test.

Comment (d-e-s-o) [reply tests/end-to-end.rs]: It sounds as if `Index` should be performing the conversion to absolute path internally then and work with that, no?

Comment (twistedfall) [reply tests/end-to-end.rs]: For that to work we still need to change the current directory into the temporary area before calling `fs::canonicalize()`, otherwise it will resolve to a directory under project root instead of temp dir.

Comment (d-e-s-o) [reply tests/end-to-end.rs]: Okay, I guess it's a circular problem if we want to test relative path usage. In that case, I think we may have to bring out the big guns and run tests in their own dedicated processes. Conceptually, [`test-fork`](https://crates.io/crates/test-fork) helps with that, but it doesn't currently support `async` tests. So I will have to check whether that's something that can be implemented.

Comment (twistedfall) [reply tests/end-to-end.rs]: Well, technically you don't need inherent support for async. You can always start a reactor manually and `spawn()` an async task.

What should we do with the PR for now? If the main logic if fine for you I would prefer to merge it with this test dropped/commented.

Comment (d-e-s-o) [reply tests/end-to-end.rs]: > For that to work we still need to change the current directory into the temporary area before calling `fs::canonicalize()`, otherwise it will resolve to a directory under project root instead of temp dir.

Yeah, I realized that as well. But I still think we should be able to pass in a suitably crafted relative path. E.g., if `cargo-http-registry` is at `/home/user/cargo-http-registry` and `tempdir` creates `/tmp/.tmpYb5HBA/`, the relative path should be `../../../tmp/.tmpYb5HBA/`. I would guess that would be sufficient for testing? Or do we specifically need something without parent directory involvement?

(that may not work for Windows, but that's acceptable)

Comment (twistedfall) [reply tests/end-to-end.rs]: I though about this, but was hesitating to bring another dep (e.g. `pathdiff`). Additionally I think it won't always conceptually work on Windows if temp dir and project dir are on the different drives (e.g. C: and D:).

Comment (d-e-s-o) [reply tests/end-to-end.rs]: Yes, I understand the Windows restriction, but I think it's fine given that it's only for testing. Same for the dependency: not a problem if it's a dev-only thing (had no idea this functionality was *not* provided by the standard library though, ugh...).

I think the relative-path avenue would be my preference, though spawning an async runtime in each forked test is fine as well.

That being said, none of the tests fail if I revert the changes to `src/index.rs`, so I don't know what they test, but it ain't what you set out to fix, I guess...

Comment (twistedfall) [reply tests/end-to-end.rs]: The changes in `index.rs` don't really affect anything now that I changed all invocations of `Index::add` to only use relative paths. If you roll back both `index.rs` and `publish.rs` changes you'll get failure in `publish_relative_index_root()` test.

Comment (twistedfall) [reply tests/end-to-end.rs]: The test now uses `pathdiff` to calculate the relative path without setting the current working directory.

File: src/publish.rs [hunk 4]
```diff
@@ -296,14 +297,14 @@ pub fn publish_crate(mut body: Bytes, index: &mut Index) -> Result<()> {
...
+  index.add(&crate_meta_relative_path).with_context(|| {
     format!(
       "failed to add {} to git repository",
       crate_meta_path.display()
     )
   })?;
   index
-    .add(&crate_path)
+    .add(&crate_relative_path)
     .with_context(|| format!("failed to add {} to git repository", crate_path.display()))?;
```
Comment (d-e-s-o) [inline src/publish.rs]:
```suggestion
    .with_context(|| format!("failed to add {} to git repository", crate_relative_path.display()))?;
```

File: src/publish.rs [hunk 5]
```diff
@@ -296,14 +297,14 @@ pub fn publish_crate(mut body: Bytes, index: &mut Index) -> Result<()> {
     .write(&data)
     .with_context(|| format!("failed to write to crate file {}", crate_path.display()))?;
 
-  index.add(&crate_meta_path).with_context(|| {
+  index.add(&crate_meta_relative_path).with_context(|| {
     format!(
       "failed to add {} to git repository",
       crate_meta_path.display()
```
Comment (d-e-s-o) [inline src/publish.rs]:
```suggestion
      crate_meta_relative_path.display()
```

File: src/publish.rs [hunk 6]
```diff
@@ -254,18 +254,19 @@ pub fn publish_crate(mut body: Bytes, index: &mut Index) -> Result<()> {
...
+  let crate_meta_relative_dir = crate_path(&crate_name);
+  let crate_meta_dir = index.root().join(&crate_meta_relative_dir);
   create_dir_all(&crate_meta_dir)
     .with_context(|| format!("failed to create directory {}", crate_meta_dir.display()))?;
 
   let crate_length = parse_u32(&mut body)
     .context("failed to read crate length")?
-    .try_into()
-    .unwrap();
+    .try_into()?;
```
Comment (d-e-s-o) [inline src/publish.rs]: Let's revert these changes, please. It's inconsistent with code higher up that still unwraps and on top of that any context is missing, so if we see such an error it will be less than useful.

**Review by d-e-s-o — APPROVED** [review]
Seems fine to me now, thanks! Left a few more minor comments. Also, can you please squash into a single commit?

File: src/publish.rs (hunk 4 above)
Comment (twistedfall) [reply src/publish.rs]: Well, I specifically left it so because it would be better to see the full path in the error message for debugging. Maybe display both?

**Comment by twistedfall** [comment]
Squashed and tended to the comments. I've added both full and relative path to the context for now, but I can remove the full path if you think it shouldn't be there.

**Comment by d-e-s-o** [comment]
Seems fine to me now, thanks!
